File: contrib/matlab/postprocess.m
% Reads binary files in a directory and does
% postprocessing (e.g., to visualize later with
% VAPOR). Note that for big runs, GHOST can do some
% automatic postprocessing (e.g., compute vorticity)
% at run time.
% Type 'postprocess.m' in MATLAB to execute

% Path to the binary data
path = '../../3D/bin/outs/';

% Spatial resolution
N = 128;
dx = 2*pi/N;

% Reads binary files, computes vertical vorticity, 
% and saves in a new binary file named 'wz.NNNN.out'
filelist = dir(strcat(path,'vx.*.out'));
nfiles = numel(filelist);
for i = 1:nfiles,
  fid = fopen(strcat(path,filelist(i).name));
  vx = fread(fid,N*N*N,'float32');
  fclose(fid);
  vx = reshape(vx,[N N N]);
  ind = filelist(i).name(4:7);
  str = ['vy.' ind '.out'];
  fid = fopen(strcat(path,str));
  vy = fread(fid,N*N*N,'float32');
  fclose(fid);
  vy = reshape(vy,[N N N]);
  adv = circshift(vy,[-1 0 0]);
  ret = circshift(vy,[1 0 0]);
  vy = (adv-ret)/(2*dx); % dv_y/dx
  adv = circshift(vx,[0 -1 0]);
  ret = circshift(vx,[0 1 0]);
  vx = (adv-ret)/(2*dx); % dv_x/dy 
  wz = vy-vx;
  str = ['wz.' ind '.out'];
  fid = fopen(strcat(path,str),'w');
  fwrite(fid,wz,'float32');
  fclose(fid);
end

File: contrib/python/postprocess.py
import numpy as np
import glob as glob
import string as string
import matplotlib.pyplot as plt

# Reads binary files in a directory and does
# postprocessing (e.g., to visualize later with
# VAPOR). Note that for big runs, GHOST can do some
# automatic postprocessing (e.g., compute vorticity)
# at run time.
# Execute in ipython with '%run postprocess.py'

# Path to the binary data
path = '../../3D/bin/outs/'

# Box size
Lx = 2*np.pi
Ly = 2*np.pi

# Spatial resolution
NX = 128
NY = 128
NZ = 128
dx = Lx/NX
dy = Ly/NX
shape = (NX,NY,NZ)

# Reads binary files, computes vertical vorticity
# using centered finite differences, and saves in 
# a new binary file named 'wz.NNNN.out'
filelist = sorted(glob.glob(path+'vx.*.out'))
for file in filelist:
  vx = np.fromfile(file,dtype=np.float32).reshape(shape,order='F')
  ind = string.lstrip(file,path+'vx.').rstrip('.out')
  str = 'vy.'+ind+'.out'
  vy = np.fromfile(path+str,dtype=np.float32).reshape(shape,order='F')
  adv = np.roll(vy,-1,axis=0)
  ret = np.roll(vy,1,axis=0)
  vy = (adv-ret)/(2*dx) # dv_y/dx
  adv = np.roll(vx,-1,axis=1)
  ret = np.roll(vx,1,axis=1)
  vx = (adv-ret)/(2*dy) # dv_x/dy 
  wz = vy-vx
  wz.tofile(path+'wz.'+ind+'.out')
